File: metropolis.py
import matplotlib.pyplot as plt
import math
import random

size = 100

T = 1.5 # temperature * boltzmann_k / epsilon

spin = [[0 for _ in range(size)] for _ in range(size)]

def initialize():
    for i in range(size):
        for j in range(size):
            spin[i][j] = random.choice([-1,1])
            # spin[i][j] = 1

def step():
    i = random.randint(0,size-1)
    j = random.randint(0,size-1)

    deltaE = energy_diff(i,j)
    if deltaE <= 0:
        spin[i][j] *= -1
    else:
        flip_probability = math.exp(-deltaE/T)
        if random.random() < flip_probability:
            spin[i][j] *= -1

def energy_diff(i, j):
    top = spin[(i-1)%size][j]
    bottom = spin[(i+1)%size][j]
    left = spin[i][(j-1)%size]
    right = spin[i][(j+1)%size]

    return 2*spin[i][j]*(top+bottom+left+right)


initialize()
for _ in range(100*size**2):
    step()
plt.imshow(spin, cmap="gray")
plt.show()

File: metropolis.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define SIZE 300
#define STEPS (SIZE*SIZE*100)

double T = 1.5; // T * k / J
char spin[SIZE][SIZE];

char energyDiff(int i, int j)
{
    char top = spin[(i-1)%SIZE][j];
    char bottom = spin[(i+1)%SIZE][j];
    char left = spin[i][(j-1)%SIZE];
    char right = spin[i][(j+1)%SIZE];

    return 2*spin[i][j]*(top+bottom+left+right);
}

double rand01()
{
    return (double)rand() / RAND_MAX;
}

// random integer in [0, n-1]
int randn(int n)
{
    return (int) (rand01()*SIZE);
}

void step()
{
    int i = randn(SIZE);
    int j = randn(SIZE);
    
    double deltaE = energyDiff(i,j);
    if (deltaE <= 0) {
        spin[i][j] *= -1;
    } else {
        double flip_probability = exp(-deltaE/T);
        if (rand01() < flip_probability)
            spin[i][j] *= -1; 
    }
}

void initialize()
{
    srand(time(NULL));
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            spin[i][j] = rand01() < 0.5 ? -1 : 1;
        }
    }
}

void export(char* filepath)
{
    FILE* fp = fopen(filepath, "w");
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            fprintf(fp, "%d,", spin[i][j] > 0 ? 1 : 0);
        }
        fprintf(fp, "\n");
    }
}

int main(int argc, char* argv[]) 
{
    printf("simulating...\n");
    initialize();
    for (int i = 0; i < STEPS; i++)
        step();
    printf("exporting...\n");
    char *dataFilepath = argv[1];
    export(dataFilepath);
    printf("done!\n");
}